File: apps/mobile/ios/Sources/Flagship/AcmeAccountKey.swift
import Foundation
import CryptoKit

/// #28 — ACME account-key recovery crypto.
///
/// The ACME account key is the authority to mint a user's TLS certs. It
/// is an ECDSA P-256 key (Let's Encrypt account keys are ES256, NOT
/// Ed25519), generated on-device, held by the admin, stored exportably,
/// and ESCROWED into the existing WebAuthn-PRF recovery envelope so
/// losing every device doesn't brick cert issuance.
///
/// This enum is the pure crypto half: deriving the cross-platform
/// `accountKeyId`, and wrapping / unwrapping the raw 32-byte private
/// scalar for escrow. Storage + keygen live in `Keystore`.
///
/// The wire/derivation contract is shared verbatim with Android and the
/// TypeScript control plane — see the known-answer vector in
/// `AcmeAccountKeyTests`. Any drift here breaks recovery across surfaces.
public enum AcmeAccountKey {

    public enum AcmeAccountKeyError: Error, LocalizedError {
        case base64Decode
        case unwrapFailed(String)

        public var errorDescription: String? {
            switch self {
            case .base64Decode:         return "Couldn't base64-decode the escrowed account key."
            case .unwrapFailed(let m):  return "Couldn't decrypt the escrowed account key: \(m)"
            }
        }
    }

    /// Domain-separated HKDF salt for the account-key escrow wrap. DELIBERATELY
    /// different from the UMK wrap salt (`flagship/recovery-wrap/v1`) — the two
    /// secrets ride the same PRF-derived input keying material, so they MUST
    /// derive distinct AES keys. Mirrored on Android + the TS daemon.
    private static let escrowSalt = Data("flagship/recovery-acme-wrap/v1".utf8)

    /// The stable cross-platform identifier for an ACME account key:
    /// lowercase sha256-hex of the uncompressed SEC1 public point
    /// (0x04‖X‖Y, 65 bytes for P-256). CryptoKit's `x963Representation`
    /// returns exactly that uncompressed encoding.
    public static func accountKeyId(publicKey: P256.Signing.PublicKey) -> String {
        let digest = SHA256.hash(data: publicKey.x963Representation)
        return digest.map { String(format: "%02x", $0) }.joined()
    }

    /// Wrap the raw 32-byte private scalar for escrow. Derives an AES key
    /// via HKDF-SHA256 over the PRF secret (under `escrowSalt`), seals the
    /// scalar with AES-GCM, and returns `sealed.combined` (nonce‖ct‖tag) as
    /// a single self-contained base64 blob.
    public static func wrapForEscrow(scalar: Data, prfSecret: Data) throws -> String {
        let key = HKDF<SHA256>.deriveKey(
            inputKeyMaterial: SymmetricKey(data: prfSecret),
            salt: escrowSalt,
            info: Data(),
            outputByteCount: 32
        )
        let sealed = try AES.GCM.seal(scalar, using: key)
        guard let combined = sealed.combined else {
            throw AcmeAccountKeyError.unwrapFailed("no combined ciphertext")
        }
        return combined.base64EncodedString()
    }

    /// Reverse of `wrapForEscrow`. Base64-decodes the combined blob,
    /// reconstructs the sealed box, and opens it under the PRF-derived key.
    /// Returns the recovered 32-byte private scalar.
    public static func unwrapFromEscrow(base64: String, prfSecret: Data) throws -> Data {
        guard let combined = Data(base64Encoded: base64) else {
            throw AcmeAccountKeyError.base64Decode
        }
        let key = HKDF<SHA256>.deriveKey(
            inputKeyMaterial: SymmetricKey(data: prfSecret),
            salt: escrowSalt,
            info: Data(),
            outputByteCount: 32
        )
        do {
            let box = try AES.GCM.SealedBox(combined: combined)
            return try AES.GCM.open(box, using: key)
        } catch {
            throw AcmeAccountKeyError.unwrapFailed(String(describing: error))
        }
    }
}

File: apps/mobile/android/app/src/main/java/com/flagshipserver/app/core/AcmeAccountKey.kt
// #28 — the device-side of ACME account-key recovery.
//
// The ACME account key is the authority to mint a user's TLS certs. Unlike
// the IRK (which is HKDF-derived from the UMK), the account key is an
// independent admin-held ECDSA P-256 (ES256) key — Let's Encrypt account
// keys are ES256, NOT Ed25519. Losing every admin device would otherwise
// brick cert issuance forever, so the raw 32-byte scalar is ESCROWED
// (wrapped) into the SAME WebAuthn-PRF recovery envelope that carries the
// UMK, under its own HKDF salt for domain separation.
//
// CROSS-PLATFORM CONTRACT (must match iOS + the .com cloud byte-for-byte):
//   - private value      = 32-byte big-endian scalar S
//   - public encoding     = uncompressed SEC1: 0x04 ‖ X(32) ‖ Y(32) (65 bytes)
//   - accountKeyId        = lowercase sha256-hex of the uncompressed pubkey
//   - escrow wrap salt    = "flagship/recovery-acme-wrap/v1" (≠ the UMK's
//                           "flagship/recovery-wrap/v1" — domain separation)
//   - escrow blob          = base64(nonce(12) ‖ ciphertext ‖ gcmTag(16))
//
// The public point is derived with BouncyCastle's secp256r1 curve math so
// publicUncompressed(scalar) works for ANY scalar (incl. the scalar=2
// known-answer vector) independent of how the key was minted.

package com.flagshipserver.app.core

import java.math.BigInteger
import java.security.KeyPairGenerator
import java.security.MessageDigest
import java.security.SecureRandom
import java.security.interfaces.ECPrivateKey
import java.security.spec.ECGenParameterSpec
import javax.crypto.Cipher
import javax.crypto.Mac
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec
import org.bouncycastle.jce.ECNamedCurveTable
import org.bouncycastle.jce.spec.ECNamedCurveParameterSpec

object AcmeAccountKey {
    /** HKDF salt for the escrow wrap. DIFFERENT from Recovery.wrap's UMK
     *  salt so the same PRF secret produces independent wrap keys. */
    val ESCROW_WRAP_SALT: ByteArray = "flagship/recovery-acme-wrap/v1".toByteArray()

    private const val CURVE = "secp256r1"
    private val spec: ECNamedCurveParameterSpec = ECNamedCurveTable.getParameterSpec(CURVE)
    private val rng = SecureRandom()

    /** Generate a fresh ECDSA P-256 private scalar (32 bytes, big-endian).
     *  Uses java.security keygen (StrongBox-independent — this key MUST be
     *  exportable for escrow) and reduces to the canonical 32-byte form. */
    fun generateScalar(): ByteArray {
        val gen = KeyPairGenerator.getInstance("EC")
        gen.initialize(ECGenParameterSpec(CURVE), rng)
        val priv = gen.generateKeyPair().private as ECPrivateKey
        return toFixed32(priv.s)
    }

    /** Uncompressed SEC1 public key (0x04 ‖ X ‖ Y, 65 bytes) for [scalar]·G
     *  on secp256r1. */
    fun publicUncompressed(scalar: ByteArray): ByteArray {
        val s = BigInteger(1, scalar)
        // `multiply` then `normalize` so getAffineX/Y are the affine
        // coordinates (BouncyCastle stores points projectively).
        val point = spec.g.multiply(s).normalize()
        val x = toFixed32(point.affineXCoord.toBigInteger())
        val y = toFixed32(point.affineYCoord.toBigInteger())
        val out = ByteArray(65)
        out[0] = 0x04
        System.arraycopy(x, 0, out, 1, 32)
        System.arraycopy(y, 0, out, 33, 32)
        return out
    }

    /** Lowercase sha256-hex of the uncompressed public key. The stable,
     *  cross-platform identifier for an account key. */
    fun accountKeyId(scalar: ByteArray): String {
        val digest = MessageDigest.getInstance("SHA-256").digest(publicUncompressed(scalar))
        return HexUtil.encode(digest)
    }

    /** Wrap the raw 32-byte [scalar] under a PRF-derived AES-256-GCM key.
     *  Returns a single self-contained base64 blob = nonce(12) ‖ ct ‖
     *  tag(16) ready to ship as `wrappedAcmeAccountKey`. */
    fun wrapForEscrow(scalar: ByteArray, prfSecret: ByteArray): String {
        val key = hkdfSha256(ikm = prfSecret, salt = ESCROW_WRAP_SALT, info = ByteArray(0), length = 32)
        val nonce = ByteArray(12).also(rng::nextBytes)
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.ENCRYPT_MODE, SecretKeySpec(key, "AES"), GCMParameterSpec(128, nonce))
        val ctWithTag = cipher.doFinal(scalar) // ciphertext ‖ 16-byte GCM tag
        val blob = ByteArray(nonce.size + ctWithTag.size)
        System.arraycopy(nonce, 0, blob, 0, nonce.size)
        System.arraycopy(ctWithTag, 0, blob, nonce.size, ctWithTag.size)
        return java.util.Base64.getEncoder().encodeToString(blob)
    }

    /** Reverse of [wrapForEscrow]: recover the 32-byte scalar. Throws
     *  AEADBadTagException on a wrong PRF secret. */
    fun unwrapFromEscrow(base64Blob: String, prfSecret: ByteArray): ByteArray {
        val blob = java.util.Base64.getDecoder().decode(base64Blob)
        require(blob.size > 12) { "escrow blob too short" }
        val nonce = blob.copyOfRange(0, 12)
        val ctWithTag = blob.copyOfRange(12, blob.size)
        val key = hkdfSha256(ikm = prfSecret, salt = ESCROW_WRAP_SALT, info = ByteArray(0), length = 32)
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.DECRYPT_MODE, SecretKeySpec(key, "AES"), GCMParameterSpec(128, nonce))
        return cipher.doFinal(ctWithTag)
    }

    /** Left-pad / strip a BigInteger to exactly 32 big-endian bytes.
     *  BigInteger.toByteArray() may add a leading 0x00 sign byte or emit
     *  fewer than 32 bytes; both are normalized here. */
    private fun toFixed32(v: BigInteger): ByteArray {
        val raw = v.toByteArray()
        val out = ByteArray(32)
        when {
            raw.size == 32 -> System.arraycopy(raw, 0, out, 0, 32)
            raw.size < 32 -> System.arraycopy(raw, 0, out, 32 - raw.size, raw.size)
            else -> System.arraycopy(raw, raw.size - 32, out, 0, 32) // drop leading sign byte(s)
        }
        return out
    }

    /** HKDF-SHA256. Identical construction to Recovery.hkdfSha256 +
     *  Keystore.hkdf — kept local so this object has no cross-package
     *  private dependency. */
    private fun hkdfSha256(ikm: ByteArray, salt: ByteArray, info: ByteArray, length: Int): ByteArray {
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(SecretKeySpec(salt, "HmacSHA256"))
        val prk = mac.doFinal(ikm)
        mac.init(SecretKeySpec(prk, "HmacSHA256"))
        val out = ByteArray(length)
        var t = ByteArray(0)
        var counter = 1
        var written = 0
        while (written < length) {
            mac.reset()
            mac.update(t)
            mac.update(info)
            mac.update(counter.toByte())
            t = mac.doFinal()
            val n = minOf(t.size, length - written)
            System.arraycopy(t, 0, out, written, n)
            written += n
            counter++
        }
        return out
    }
}
